### PR TITLE
refactor: support image in svg and prefetch external links in element

### PR DESCRIPTION
**Pull request by Jackie1210**
Close https://github.com/vercel/satori/issues/312
Close https://github.com/vercel/satori/issues/468

## Refactor
currently I just collect all links before `layout` and resolveImageData

## Enhancemant
make image work in svg

## Additional
Some test broke because of recursive import between `handle/image` and `utils`, so I just put them in one file `preprocess`

You can preview [it](https://satori-playground-git-fork-jackie1210-image.vercel.sh/?share=dVPtbpswFH0V5GniTxQbSNiCQqWt3bTPP2ulStMkRPDFODU2wyaQRnn32Wm3wtIiIXHPObo-5_pyQIWigBK0pnz3S3qPjzZ7Aenh8IR4XgWcVSbx_ICQ1_5sTPWcmupZhnLdiHxvuVLAMOUccsVbKAxX0ioKJbpaTjW54Ex-NlBrJwBpoJ0Ktp02vNxfKktJ5-450SYv7lirOkkvlVCtVb0qy_I_N7bDNb-HxIvCM-L2MXtMyIg7Hv9-XzyBa71jZ7NJ3Wj882me4zsO_Xs1pD7xiGdJ904EJRci9aWSMIGHWkid-pUxTYJx3_fzPpqrluGQEIKtp5F65Nb65XXOYIx4Hqepf4JJFmTh5JwXA70cyTItlA_etDVnonlpCirnEgzeNgyTEK-WOIzxIsZRTLKPWbhahvEifrPMCtF__Tko9q4avl2v9FUQRzfs7j64_f5J_w5-fJnbBpPT8PguXPB_9RrbHb9AM6Qat3EaJQd0CoOSt_Ze0YN_lCxcQWHTMZSUudAwQ1CrLb_ZN-5PMf2psn3cZnyoN0BRYtoOjjNk8o1VVCCE6lUrKDr-AQ)

### CODE REVIEW

**Comment by LuciNyan** [comment]
🤯 You rocked it!!!

File: test/image.test.tsx [hunk 1]
```diff
@@ -103,6 +103,45 @@ describe('Image', () => {
...
+          />
+        </svg>
+      </div>,
+      { width: 100, height: 100, fonts }
+    )
+    expect(toImage(svg, 100)).toMatchImageSnapshot()
+
+    expect(requests).toEqual([
+      'https://t3.ftcdn.net/jpg/02/95/26/46/360_F_295264675_clwKZxogAhxLS9sD163Tgkz1WMHsq1RJ.jpg',
+    ])
```
Comment (shuding) [inline test/image.test.tsx:142]: We need to use base64 encoded image to avoid the URL being broken.

Comment (Jackie1210) [reply test/image.test.tsx:142]: @shuding, here I just follow the following steps
- fetch the image
- transform it to base64 encoded image data and store it in k-v map
- when we meet `svg`, we just do `SVGNodeToImage` and replace absolute url to base64 data in store.

Is there anything I didn't understand?😷

Comment (Jackie1210) [reply test/image.test.tsx:142]: Or did you mean we just use a base64 format image in this test case? But we do need a absolute link here. So I think I should change an absolute link would never broken instead?

Comment (Jackie1210) [reply test/image.test.tsx]: @shuding done

File: src/handler/preprocess.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    })
+    .join('')}>${translateSVGNodeToSVGString(children, currentColor)}</${type}>`
+}
+/**
+ * pre process node and resolve absolute link to img data for image element
+ * @param node ReactNode
+ * @returns
+ */
+export async function preProcessNode(node: ReactNode) {
+  const set = new Set<string>()
```
Comment (shuding) [inline src/handler/preprocess.ts:137]:
```suggestion
  const set = new Set<string | Buffer | ArrayBuffer>()
```

File: src/satori.ts [hunk 3]
```diff
@@ -87,6 +88,8 @@ export default async function satori(
   // To address this situation, we may need to add `processedWordsMissingFont`
   const processedWordsMissingFonts = new Set()
```
Comment (shuding) [inline src/satori.ts:90]: We probably need to import the LRU cache from `image.ts` and clear it before preprocessing. I think I've seen reports about image being cached across multiple `satori()` calls.

File: src/handler/preprocess.ts (hunk 2 above)
Comment (Jackie1210) [reply src/handler/preprocess.ts]: @shuding done

**Review by shuding — APPROVED** [review]
Thanks!

**Pull request merged**